### PR TITLE
fix ndt score function naming: likelihood instead of probability

### PR DESCRIPTION
**Pull request by grischi**
The score function of NDT as implemented based on [Magnusson 2009], Eq. 6.9 & 6.10 is a likelihood, not a probability. This may sound like an irrelevant naming detail, but may cause confusion for example when obtaining values larger than 1 from getTransformationProbability. This PR replaces "probability" with "likelihood".

### CODE REVIEW

**Comment by mvieth** [comment]
Thank you for opening this pull request. Unfortunately, it isn't as easy as simply renaming everything. Changing the documentation is no problem, but `getTransformationProbability` is a public function and `trans_probability_` is a protected variable (may be used in a subclass), so simply renaming these will create problems for everyone who uses them. If you want to rename `getTransformationProbability`, you have to create a new function with the new name, and add a deprecation notice to the old function. Renaming the variable is a bit more difficult, there would have to be two instances of the variable, one with the old and one with the new name, the old one with a deprecation notice, and you would have to make sure that both variables always have the same value (not sure how worth this is, maybe a strong hint in the documentation that it is actually the likelihood would suffice?)

**Comment by grischi** [comment]
Thank you for your comment and sorry for forgetting these issues with interface compatibility. I restored getTransformationProbability and put a deprecation note there. I also restored trans_probability_ as a reference to trans_likelihood_, also with a deprecation note.
I can't tell if it is worth it, however, i'm quite sure that a note in the documentation telling that it is called probability but actually is a likelihood is even more confusing. I believe so because [Magnusson 2009] approximates a likelihood function using a Gaussian (this score function) that obviously really "mathematically looks like" a probability density function.

**Comment by grischi** [comment]
I had to remove the deprecation macro from trans_probability_ since it was not able to build on ubuntu 18.04 gcc.

**Comment by mvieth** [comment]
Having both a variable and a reference to it is also quite confusing and not desirable in the long term. Have you tried if it works without the explicit initialization of `trans_likelihood_` in the constructor, or does simply assigning the reference of `trans_likelihood_` to `trans_probability_` cause the deprecation warning?
Another solution might be a union that contains both variables, the old one deprecated. But technically, only one member of the union can be active at a time, and reading from the member that was not written to last would be undefined behaviour, although many compilers support that

**Comment by mvieth** [comment]
Hmm, the GCC diagnostic pragma makes the build pass, but it doesn't seem very elegant to me and MSVC prints warnings that it does not know the pragma. As a side note, it is kind of weird that only GCC prints the deprecation warning when initializing the deprecated reference, while Clang and MSVC don't.
Would you test the solution with a union I suggested above? I verified that at least GCC and Clang support reading from the passive member, and I am pretty sure that MSVC does too. And this is only a temporary solution, so it should be fine

**Comment by grischi** [comment]
I agree, pragmas are not elegant. I have added some exclusions for msvc and clang that are not elegant either, but at least should not produce compiler warnings.
Regarding the union I was not able to keep interface compatiblity and at the same time not produce a compiler warning. Interface compatibility requires a reference to the deprecated union member, which again causes a warning. Or did i misunderstand something here?

**Comment by mvieth** [comment]
> Regarding the union I was not able to keep interface compatiblity and at the same time not produce a compiler warning. Interface compatibility requires a reference to the deprecated union member, which again causes a warning. Or did i misunderstand something here?

I had something like this in mind (but I didn't test it extensively):
```c++
  union {
    PCL_DEPRECATED(1, 16, "TODO message")
    double trans_probability_;
    double trans_likelihood_;
  };
```

**Comment by grischi** [comment]
Sure, but how would you assign to `trans_probability_` (like in `impl/ndt.hpp:143`) without causing a warning?

File: registration/include/pcl/registration/ndt.h [hunk 1]
```diff
@@ -168,13 +168,26 @@ class NormalDistributionsTransform : public Registration<PointSource, PointTarge
...
+
   /** \brief Get the registration alignment probability.
    * \return transformation probability
    */
+  PCL_DEPRECATED(1,
+                 15,
+                 "The method `getTransformationProbability` is deprecated. Please use "
+                 "`getTransformationLikelihood` instead.")
   inline double
   getTransformationProbability() const
```
Comment (mvieth) [inline registration/include/pcl/registration/ndt.h:188]: Yes, I think that's a good solution with the union. Just two minor things: please deprecate this function for 1.16, not 1.15, and make it clear in the documentation that this is just a renaming, that the returned value is the same as before (getTransformationLikelihood vs getTransformationProbability)

Comment (grischi) [reply registration/include/pcl/registration/ndt.h:188]: I agree, didn't think of an anonymous union on first try. By documentation, do you mean the deprecation message or the function comment \brief?

Comment (mvieth) [reply registration/include/pcl/registration/ndt.h:188]: Yep, the new deprecation message with the word "renamed" should do it

**Review by mvieth — APPROVED** [review]

File: registration/include/pcl/registration/ndt.h [hunk 2]
```diff
@@ -168,13 +168,26 @@ class NormalDistributionsTransform : public Registration<PointSource, PointTarge
...
+    return trans_likelihood_;
+  }
+
   /** \brief Get the registration alignment probability.
    * \return transformation probability
    */
+  PCL_DEPRECATED(1,
+                 16,
+                 "The method `getTransformationProbability` was renamed to "
+                 "`getTransformationLikelihood`.")
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/ndt.h:186]: Would present tense make more sense?

**Comment by kunaltyagi** [comment]
Not this PR, but the documentation blocks in this file need to be fixed

```
   * vector. \note Equation 6.10, 6.12 and 6.13 [Magnusson 2009]. \param[out]
   * score_gradient the gradient vector of the likelihood function w.r.t. the
   * transformation vector \param[out] hessian the hessian matrix of the likelihood
   * function w.r.t. the transformation vector \param[in] trans_cloud transformed point
   * cloud \param[in] transform the current transform vector \param[in] compute_hessian
   * flag to calculate hessian, unnessissary for step calculation.

```
								
[registration/include/pcl/registration/ndt.h](https://github.com/pointcloudlibrary/pcl/blob/5a759106fa33c7881f2cb76920025cf6c1591e94/registration/include/pcl/registration/ndt.h#L274-L280) (Lines 274-280)

File: registration/include/pcl/registration/ndt.h [hunk 3]
```diff
@@ -483,9 +496,16 @@ class NormalDistributionsTransform : public Registration<PointSource, PointTarge
...
-  /** \brief The probability score of the transform applied to the input cloud,
+  /** \brief The likelihood score of the transform applied to the input cloud,
    * Equation 6.9 and 6.10 [Magnusson 2009]. */
-  double trans_probability_;
+  union {
+    PCL_DEPRECATED(1,
+                   16,
+                   "The member `trans_probability_` is deprecated. Please use "
+                   "`trans_likelihood_` instead.")
+    double trans_probability_;
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/ndt.h:506]: PCL_DEPRECATED add some of the message in the macro. Maybe reword it slightly?

**Review by kunaltyagi — COMMENTED** [review]
Just minor comments. Use of union was a very good idea

**Comment by grischi** [comment]
> Not this PR, but the documentation blocks in this file need to be fixed
> 
> ```
>    * vector. \note Equation 6.10, 6.12 and 6.13 [Magnusson 2009]. \param[out]
>    * score_gradient the gradient vector of the likelihood function w.r.t. the
>    * transformation vector \param[out] hessian the hessian matrix of the likelihood
>    * function w.r.t. the transformation vector \param[in] trans_cloud transformed point
>    * cloud \param[in] transform the current transform vector \param[in] compute_hessian
>    * flag to calculate hessian, unnessissary for step calculation.
> ```
> 
> [registration/include/pcl/registration/ndt.h](https://github.com/pointcloudlibrary/pcl/blob/5a759106fa33c7881f2cb76920025cf6c1591e94/registration/include/pcl/registration/ndt.h#L274-L280) (Lines 274-280)

Thank you for your review. If you mean the newlines, there are some more comments where these are missing in this file. I assume with "Not this PR" you mean that these newline errors were not introduced in this PR. Fixing all these other comments within this PR will however be confusing as many of the other comments with missing newlines do not mention probability/likelihood, right?

File: registration/include/pcl/registration/ndt.h (hunk 2 above)
Comment (grischi) [reply registration/include/pcl/registration/ndt.h:186]: Please correct me if I'm wrong, I understand it like this: The method _was_ renamed and therefore it _is_ now deprecated.

File: registration/include/pcl/registration/ndt.h (hunk 3 above)
Comment (grischi) [reply registration/include/pcl/registration/ndt.h:506]: I agree, will push another commit when the other questions are resolved.

**Comment by kunaltyagi** [comment]
> Should I fix them within this PR anyway?

No
